File: uva/10263 Railway/railway.py
def nearestPoint(m: complex, p1: complex, p2: complex) -> complex:
    """ Find the point on the line A closest to point m
        A is the line determined by points p1 and p2 """
    # Do a ternary search
    hi, lo = p2, p1
    while abs(hi - lo) > 1e-8:
        m1 = lo + (hi - lo) * 0.3
        m2 = lo + (hi - lo) * 0.6
        if abs(m1 - m) > abs(m2 - m):
            lo = m1
        else:
            hi = m2
    return lo

xm = float(input())
while True:
    ym = float(input())
    m = xm + ym * 1j
    # print(m)
    n = int(input())
    points = []
    for _ in range(n + 1):
        x, y = float(input()), float(input())
        points.append(x + y * 1j)
    
    ans = 0 + 0j
    dist = 10000000
    for p1, p2 in zip(points, points[1:]):
        nearest = nearestPoint(m, p1, p2)
        if abs(nearest - m) < dist:
            dist = abs(nearest - m)
            ans = nearest
    
    print("{:.4f}".format(ans.real))
    print("{:.4f}".format(ans.imag))
    
    try:
        xm = float(input())
    except:
        break

File: uva/10263 Railway/railway.cpp
#include <bits/stdc++.h>

using namespace std;

struct point{
    double x, y;
    point(double _x, double _y) : x(_x), y(_y) {}
};

struct vec {
    double x, y; // name: ‘vec’ is different from STL vector
    vec(double _x, double _y) : x(_x), y(_y) {}
};

vec toVec(point a, point b) { // convert 2 points to vector a->b
    return vec(b.x - a.x, b.y - a.y);
}

vec scale(vec v, double s) { // nonnegative s = [<1 .. 1 .. >1]
    return vec(v.x * s, v.y * s); // shorter.same.longer
}

point translate(point p, vec v) { // translate p according to v
    return point(p.x + v.x , p.y + v.y);
}

double dist(point p1, point p2) { // Euclidean distance
    // hypot(dx, dy) returns sqrt(dx * dx + dy * dy)
    return hypot(p1.x - p2.x, p1.y - p2.y); // return double  
}

double dot(vec a, vec b) { return (a.x * b.x + a.y * b.y); }
double norm_sq(vec v) { return v.x * v.x + v.y * v.y; }

// returns the distance from p to the line defined by
// two points a and b (a and b must be different)
// the closest point is stored in the 4th parameter (byref)
double distToLine(point p, point a, point b, point &c) {
    // formula: c = a + u * ab
    vec ap = toVec(a, p), ab = toVec(a, b);
    double u = dot(ap, ab) / norm_sq(ab);
    c = translate(a, scale(ab, u)); // translate a to c
    return dist(p, c); // Euclidean distance between p and c
}

int main() {
    double x, y;
    while (cin >> x >> y) {
        point m({x, y});
        int n;
        cin >> n;
        vector<point> points;
        
        for (int i = 0; i <= n; i ++) {
            cin >> x >> y;
            points.push_back({x,y});
        }
        double dist = 10000000000.0;
        point ans(0,0);
        for (int i = 0; i <= n; i ++) {
            point p = points[i];
            point p2 = points[(i + 1) % (n + 1)];
            point c(0,0);
            double tmp = distToLine(m, p, p2, c);
            // cout << "(" << p.x << " " << p.y << "), (" << p2.x << " " << p2.y << "), (" << c.x << " " << c.y << endl;
            if (tmp < dist) {
                dist = tmp;
                ans = c;
            }
        }
        cout << setprecision(4) << fixed << ans.x << endl << ans.y << endl;
    }
}